torch/_inductor/codegen/custom_extern_kernel_codegen.py

```diff
@@ -0,0 +1,109 @@
+# mypy: allow-untyped-defs
+"""
+Custom extern kernel codegen registry.
+
+This module provides a registry that maps operators to their custom codegen
+implementations. This allows us to keep all custom implementations in one place
+and easily extend support for both Python and C++ wrappers.
+
+To add a new custom implementation:
+1. Create a codegen function with signature:
+   def my_codegen(node: ir.FallbackKernel, writeline: Callable[[str], None]) -> None:
+2. Register it in CUSTOM_EXTERN_KERNEL_CODEGEN with the operator name as key
+
+Example:
+    CUSTOM_EXTERN_KERNEL_CODEGEN = {
+        "torch.ops.higher_order.print": {
+            "python": generate_print_python,
+            "cpp": generate_print_cpp,  # Optional, falls back to python if not provided
+        },
+    }
+"""
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+
+if TYPE_CHECKING:
+    from collections.abc import Callable
+
+    from .. import ir
+
+
+def generate_print_python(
+    node: ir.FallbackKernel,
+    writeline: Callable[[str], None],
+) -> None:
+    """
+    Generate a builtin print call for the print HOP fallback (Python wrapper).
+
+    This function generates Python code that calls the builtin print function
+    with format string interpolation.
+
+    Args:
+        node: The FallbackKernel IR node representing the print HOP call.
+        writeline: A function that writes a line of code to the output buffer.
+
+    Example generated code:
+        print('x = {x}, y = {y}'.format(x=buf0, y=buf1))
+    """
+    codegen_args = node.codegen_args()
+    codegen_kwargs = node.codegen_kwargs()
+
+    # First arg is the format string
+    format_str = codegen_args[0] if codegen_args else "''"
+
+    if codegen_kwargs:
+        kwargs_str = ", ".join(codegen_kwargs)
+        writeline(f"print({format_str}.format({kwargs_str}))")
+    else:
+        writeline(f"print({format_str})")
+
+
+# Registry mapping operator names to their custom codegen implementations
+# Each entry can have:
+#   - "python": codegen function for Python wrapper (required)
+#   - "cpp": codegen function for C++ wrapper (optional, falls back to default if not provided)
+CUSTOM_EXTERN_KERNEL_CODEGEN: dict[
+    str,
+    dict[str, Callable[[ir.FallbackKernel, Callable[[str], None]], None]],
+] = {
+    "torch.ops.higher_order.print": {
+        "python": generate_print_python,
```

Comment:
instead of a dictionary can you make a dataclass or named tuple? so we can directly access it, like "custom_kernel_codegen[op].python"